File: fditools/validation/residtest.py
"""Whiteness-of-residuals identification test (port of ``residtest.m``)."""

from __future__ import annotations

import numpy as np

from ._common import models_list, eval_channel


def _xcorr_unbiased(x):
    """Unbiased auto-correlation, full length ``2N-1`` (MATLAB ``xcorr(.,'unbiased')``)."""
    x = np.asarray(x)
    N = x.size
    full = np.correlate(x, x, mode="full")        # lags -N+1 .. N-1
    scale = N - np.abs(np.arange(-N + 1, N))
    return full / scale


def residtest(x, y, freq, FRF, SYS, sCR, fs):
    """Return ``(lags, corr, cb50, frac50, tag, cb95, frac95)``."""
    x = np.atleast_2d(x)
    y = np.atleast_2d(y)
    if x.shape[0] == 1:
        x = x.T
    if y.shape[0] == 1:
        y = y.T
    freq = np.asarray(freq, dtype=float).ravel()
    FRF = np.atleast_2d(FRF)
    if FRF.shape[0] != freq.size:
        FRF = FRF.T
    sCR = np.atleast_2d(sCR)
    if sCR.shape[0] != freq.size:
        sCR = sCR.T

    nroff = freq.size
    nrofi = x.shape[1]
    nrofo = y.shape[1]
    nrofh = nrofi * nrofo
    names, models = models_list(SYS)
    nrofm = len(models)

    lags = np.arange(-nroff + 1, nroff)            # length 2*nroff-1
    nrofp = x.shape[0] / fs * (freq[1] - freq[0])

    scale0 = (nrofp - 2) / (nrofp - 1)
    scale = (nrofp - 5.0 / 3.0) / (nrofp - 11.0 / 12.0)
    cb_scale0 = scale0 * ((nrofp - 1) ** 1.5 / (nrofp - 2) / (nrofp - 3) ** 0.5)
    cb_scale = scale * (nrofp - 1) / (nrofp - 2) * np.ones_like(lags, dtype=float)
    cb_scale[nroff - 1] = cb_scale0
    ac_scale = scale * np.ones_like(lags, dtype=float)
    ac_scale[nroff - 1] = scale0

    p50 = np.sqrt(-np.log(1 - 0.5))
    p95 = np.sqrt(-np.log(1 - 0.95))
    conf = cb_scale / (nroff - np.abs(lags)) ** 0.5
    cb50 = p50 * conf
    cb95 = p95 * conf

    select = np.ones(2 * nroff - 1, dtype=bool)
    select[nroff - 1] = False                      # drop zero lag

    frac50 = np.zeros((nrofm, nrofh))
    frac95 = np.zeros((nrofm, nrofh))
    corr = np.zeros((lags.size, nrofm, nrofh))
    tag = []
    for h in range(nrofh):
        i = h // nrofo
        o = h - i * nrofo
        for m in range(nrofm):
            FRFsys = eval_channel(models[m], o, i, freq)
            res = (FRF[:, h] - FRFsys) / sCR[:, h] ** 0.5
            ac = _xcorr_unbiased(res) * ac_scale
            frac50[m, h] = np.count_nonzero(
                np.abs(ac[select]) - cb50[select] > 0) / (2 * nroff - 2) * 100
            frac95[m, h] = np.count_nonzero(
                np.abs(ac[select]) - cb95[select] > 0) / (2 * nroff - 2) * 100
            corr[:, m, h] = np.abs(ac)
        order = np.argsort(-frac50[:, h])
        frac50[:, h] = frac50[order, h]
        frac95[:, h] = frac95[order, h]
        corr[:, :, h] = corr[:, order, h]
        tag.append([names[k] for k in order])
    return lags, corr, cb50, frac50, tag, cb95, frac95

File: MATLAB/src/5_SelectionValidation/residtest.m
function [lags,corr,cb50,frac50,tag,cb95,frac95] = residtest(x,y,freq,FRF,SYS,sCR,fs)
% RESID - Identification whiteness residuals test (MIMO).
%   [lags,corr,cb50,frac50,tag,cb95,frac95] = residtest(x,y,freq,FRF,SYS,sCR,fs)
% x,y,freq  : Input & output time domain data
% FRF,SYS   : FRF measurement & structured estim models
% sCR       : Cramer-Rao lower bound of measurement
% lags,corr : frequency lags & auto_corr matrix
% cb50/95   : Model residual 50/95%-confindence bounds
% frac,tag  : fraction under confidence bounds & model name
% Author    : Thomas Beauduin, KULeuven, PMA division, 2014
%%%%%
nroff = length(freq);               % number of frequency lines
nrofi = size(x,2);                  % number of inputs vectors
nrofo = size(y,2);                  % number of output vectors
nrofh = nrofi*nrofo;                % number of transfer functions
nrofm = length(fieldnames(SYS));    % number of system models
lags = (-nroff+1:nroff-1);          % whiteness frequency lags
nrofp = length(x(:,1))/fs*(freq(2)-freq(1));

% Calculation of scaling parameters
scale0 = (nrofp-2)/(nrofp-1);
scale = (nrofp-5/3)/(nrofp-11/12);
cb_scale0 = scale0*((nrofp-1)^(3/2)/(nrofp-2)/(nrofp-3)^(1/2));
cb_scale = scale*(nrofp-1)/(nrofp-2);
cb_scale = cb_scale*ones(size(lags));
cb_scale(nroff) = cb_scale0;
ac_scale = scale*ones(size(lags));
ac_scale(nroff) = scale0;

% Calculation of confidence Bounds
p50 = sqrt(-log(1-0.5)); p95 = sqrt(-log(1-0.95));
conf_bound = repmat(cb_scale./(nroff-abs(lags)).^(0.5),[2,1]);
cb50 = p50*conf_bound(1,:);
cb95 = p95*conf_bound(2,:);

% Calculation of auto-correlation & fraction
SYS_c = struct2cell(SYS);
select = (1:2*nroff-1); 
select(nroff) = [];
tag = cell(nrofm,nrofh);
frac50 = zeros(nrofm,nrofh);
frac95 = zeros(nrofm,nrofh);
corr = zeros(length(lags),nrofm,nrofh);
for h=1:nrofh
    i = ceil(h/nrofo); o = h-(i-1)*nrofo;
    for m = 1:nrofm
        FRFsys = squeeze(freqresp(SYS_c{m}(o,i),freq*2*pi));
        res = (FRF(:,h)-FRFsys)./sCR(:,h).^0.5;
        auto_corr = local_acorr_unbiased(res).*ac_scale;
        frac50(m,h) = length(find(abs(auto_corr(select)) - ...
                              cb50(select) > 0)) / (2*nroff-2)*100;
        frac95(m,h) = length(find(abs(auto_corr(select)) - ...
                              cb95(select) > 0)) / (2*nroff-2)*100;
        corr(:,m,h) = squeeze(abs(auto_corr));
    end
   [frac50(:,h),index] = sort(frac50(:,h),'descend');
    frac95(:,h) = frac95(index,h);
    corr(:,:,h) = corr(:,index,h);
    tag(:,h) = fieldnames(SYS);
    tag(:,h) = tag(index,h);
end

end

% ===== local helper (replaces xcorr -> no Signal Processing Toolbox) =========
function r = local_acorr_unbiased(v)
% Unbiased autocorrelation of vector V (FFT-based), returned as a ROW vector
% ordered for lags -(N-1):(N-1). Drop-in for xcorr(v,'unbiased') (matches to
% machine precision; only abs(.) is used downstream). No toolbox required.
v    = v(:);
N    = numel(v);
nfft = 2*N-1;                          % exact: linear == circular autocorr
a    = ifft(abs(fft(v,nfft)).^2);      % a(1)=lag0, a(k+1)=lag +k, a(nfft-k+1)=lag -k
r    = [a(N+1:nfft); a(1:N)].';        % reorder to lags -(N-1)..(N-1), as ROW
r    = r ./ (N - abs(-(N-1):(N-1)));   % 'unbiased' scaling
end
